Created URI Signing Plugin.

alficles opened the pull request:
This is a first round of a URI Signing Plugin that conforms with https://tools.ietf.org/html/draft-ietf-cdni-uri-signing-12 . It's still a WIP, but I believe it's effectively complete. I would be very interested in your feedback, especially as this is my first submission of any substance.

I feel sure there are critical things I've failed to consider, so feel free to nitpick on even the slightest things. :)

bryancall commented:
[approve ci]

zwoop commented:
Needs to run clang-format on it.

zwoop commented on plugins/Makefile.am:
```diff
@@ -76,6 +76,10 @@ include experimental/stream_editor/Makefile.inc
 include experimental/ts_lua/Makefile.inc
 include experimental/url_sig/Makefile.inc
 
+if BUILD_URI_SIGNING_PLUGIN
```
I think my preference would be of a more "generic" type of check. Check for the libraries that you need, and then build this plugin conditional on the existence of your requirements. That way, future additions are a lot more straight forward if they have the same or similar dependencies.

alficles replied:
That define is specified in `configure.ac:1283`:

    AM_CONDITIONAL([BUILD_URI_SIGNING_PLUGIN], [test ! -z "${LIBCJOSE}" -a ! -z "${LIBJANSSON}" -a "x${enable_pcre}" = "xyes" -a "x${has_libcrypto}" = "x1"])

I had to add tests for cjose, jansson, and crypto, but I was able to reuse the pcre test. If you have the required libraries, it will build the plugin. Since cjose and jansson are not usefully available via standard packagers, it will automatically statically link them if you compile them for it. (By default, everything will be dynamic. See the README for details.)

I modeled it after how the memcached plugin is handled. It's safe to say that autoconf is not my forte, though.

alficles replied:
Here's the memcached lines:

    #
    # Enable experimental/memcached_remap plugin
    #
    PKG_CHECK_MODULES([LIBMEMCACHED], [libmemcached >= 1.0], [have_libmemcached=yes], [have_libmemcached=no])
    AM_CONDITIONAL([BUILD_MEMCACHED_REMAP_PLUGIN], [test "x${have_libmemcached}" = "xyes"])

I talked with @PSUdaemon about it, and he suggested leaving the library detection in `configure.ac`, but then exporting all those variables and putting the logic on which libraries are required to build this in this `plugins/Makefile.am`. Is that what you're suggesting?

PSUdaemon replied:
If you did export "all those variables", what would that look like?

alficles commented:
@zwoop  And yup, I missed the format. I'll get that run.

bryancall commented:
[approve ci]

SolidWallOfCode commented:
Is this intended as a replacement for `url_sig`?

alficles commented:
Yes, though because they are incompatible, people are likely to need to run both for the near future.

PSUdaemon commented:
It's also likely to have breaking changes in the near future.

PSUdaemon commented:
@alficles You have conflicts with master and autest is failing. I think rebasing might fix both.

PSUdaemon commented on plugins/experimental/uri_signing/parse.c:
```diff
@@ -0,0 +1,185 @@
+  const char *query = uri;
+  const char *end   = uri + uri_ct;
+  while (query != end && *query != '?')
+    ++query;
```
Why pre-increment?

alficles replied:
Why not?

If it doesn't matter, I usually use pre-increment, just because it has no strange unexpected behaviour. It's a tiny reduction on cognitive load. Here, either works trivially.

PSUdaemon commented on plugins/experimental/uri_signing/parse.c:
```diff
@@ -0,0 +1,185 @@
+    if (jerr.text[0]) {
+      PluginDebug("Cannot parse json for %16p: %.*s '%s'", jws, (int)pt_ct, pt, jerr.text);
+    } else {
+      PluginDebug("Cannot parse jwt for %16p: %.*s", jws, (int)pt_ct, pt);
```
I like the curly braces around the single lines here. 👍

PSUdaemon commented:
Oh, you should also squash the clang-format.

alficles commented on plugins/experimental/uri_signing/Makefile.inc:
```diff
@@ -0,0 +1,28 @@
+  experimental/uri_signing/parse.c                \
+  experimental/uri_signing/timing.c
+
+experimental_uri_signing_uri_signing_la_LIBADD = @LIBJANSSON@ @LIBCJOSE@ @LIBPCRE@ -lm -lpcre -lcrypto
```
Err... good question. That is incorrect, one way or another. Will fix.

alficles commented on plugins/experimental/uri_signing/config.c:
```diff
@@ -0,0 +1,240 @@
+  config_delete(cfg);
+issuer_fail:
+  json_decref(issuer_json);
+fail:
```
I "avoid goto at modest costs". Downward goto isn't too bad and doesn't have any particularly confusing scenarios, especially in C. It's a bit dicier in C++, ofc.

alficles commented:
All right. I've fixed the stuff I said I'd fix, including the addition of about a million curly braces. O.o I've also rebased closer to master, so hopefully I've gotten both the conflict and autest failure in one step.

scw00 commented:
[approve ci]
